signac/contrib/utility.py

```diff
@@ -227,3 +228,66 @@ def _extract(filename):
                 yield tmpdir
         else:
             raise RuntimeError("Unknown file type: '{}'.".format(filename))
+
+
+def _dotted_keys_to_nested_dicts(sp, delimiter_nested='.'):
+    """Convert dotted keys in the state point dict to a nested dict.
+
+    :param sp: A mapping with dots/delimiter_nested in its keys, e.g. {'a.b': 'c'}.
+    :param delimiter_nested: A string to represent delimiter between keys.
+    :returns: A mapping instance with nested dicts, e.g. {'a': {'b': 'c'}}.
+    """
+    ret = dict()
+    for key, value in sp.items():
+        tokens = key.split(delimiter_nested)
+        if len(tokens) > 1:
+            tmp = ret.setdefault(tokens[0], dict())
+            for token in tokens[1:-1]:
+                tmp = tmp.setdefault(token, dict())
+            tmp[tokens[-1]] = value
+        else:
+            ret[tokens[0]] = value
+    return ret
+
+
+class _hashable_dict(dict):
+    def __hash__(self):
+        return hash(tuple(sorted(self.items())))
+
+
+def _to_hashable(l):
+    if type(l) is list:
+        return tuple(_to_hashable(_) for _ in l)
+    elif type(l) is dict:
+        return _hashable_dict(l)
+    else:
+        return l
+
+
+def _encode_tree(x):
+    if type(x) is list:
+        return _to_hashable(x)
+    else:
+        return x
+
+
+def _nested_dicts_to_dotted_keys(t, encode=_encode_tree, key=None):
+    """Convert dict of dict representation of nested dict to string
```

Comment:
This function does not return a dict, so its docstrings are inaccurate. It yields tuples of `dotted_key, value`. I think the name is fine, however, since it does not claim to return a `dict` in the name. (See also comment above.)